### ISSUE TITLE
Basic Question 

### ISSUE THREAD

**Issue by Drachenkaetzchen** [TrackableEntities/trackable-entities#226]
Hi,

I'm having trouble getting a very basic test to work. Unfortunately the samples are compex, and don't cover a very basic use case with trackable entities without any client-side change tracking required, just across `DbContext`. Note that I'm using EF6 CodeFirst, with the `Product` and `Category` models from the examples.

Here's the stripped down code:

```csharp
var category =  new Category {CategoryName = "Testcategory"};
var product = new Product {ProductName = "Testproduct"};

product.Category = category;

ChangeTrackingCollection<Product> products = new ChangeTrackingCollection<Product>(true);
products.Add(product);

ChangeTrackingCollection<Category> categories = new ChangeTrackingCollection<Category>(true);
categories.Add(category);

using (var context = new TrackableEntitiesTestContext(dbPath))
{
    context.ApplyChanges(categories.GetChanges());
    context.ApplyChanges(products.GetChanges());
     
    context.SaveChanges();
    products.AcceptChanges();
    categories.AcceptChanges();
}
```

In the example above, the category gets added twice. Observations:

1. If I remove `context.ApplyChanges(categories.GetChanges());` and the corresponding change tracker, no category gets persisted.
2. If I just remove `context.ApplyChanges(categories.GetChanges());`, the example works as expected

Of course, I could just stick with observation 2, but what if the user adds a new category without products and a new product within the same call?

To me, as a developer, I would have expected that the behavior would be similar to the stock EF6 behavior, where I would simply add the `Product` to a DbSet and EF6 figures out that no `Category` has been inserted, and inserts it in the database.

I'm also puzzled regarding the `CategoryChangeTracker` on the `Product` model, as it doesn't seem to be ever called by any execution path.

EDIT: I fear that I might have misunderstood the purpose of `ChangeTrackingCollection` - up until now, I assumed that `ChangeTrackingCollection` is a drop-in replacement for `List<>`, which it doesn't seem to be.

**Comment by tonysneed** [comment]
Let me take some time to replicate your scenario. But I can make some general comments which might help.

The client-side library is entirely optional. It’s only purpose is to set the TrackingState and ModifiedProperties (for partial updates) properties on each entity. But you can instead just set these manually.

Secondly, you should only have to call ApplyChanges on the root object(s). It will traverse the object graph in all directions.

I will provide a more detailed response to your questions when I have some time in the next day or two.

**Comment by Drachenkaetzchen** [comment]
> The client-side library is entirely optional. It’s only purpose is to set the TrackingState and ModifiedProperties (for partial updates) properties on each entity. But you can instead just set these manually.

Yes, I already did start on a custom "client-side" `ModifiedProperties` before I realized that there's something odd.

> Secondly, you should only have to call ApplyChanges on the root object(s). It will traverse the object graph in all directions.

So basically I have to pass all changes to `ApplyChanges` in one go, so that the tracker can go through everything at once? I'm thinking how to make that possible. Basically my (current) use case is to load everything in my WPF at startup, and syncing changes to the database when the user wishes to save.

> I will provide a more detailed response to your questions when I have some time in the next day or two.

No need to hurry - I'm currently thinking about dropping EF6 altogether and go with serialization or something. It seems that EF6 is really slow with a large number of entities. 100k is a usual number in my scenario, which I work around by only loading partial data.

That's also the reason why I don't go with CRUD or UOW, because I try to keep it as simple as possible from a developer's point of view.
